src/main/clojure/clara/rules/memory.cljc

```diff
@@ -241,7 +256,78 @@
+                     ;; pre-1.9.0.
+                     (if-let [^clojure.lang.ISeq entries (.seq ^clojure.lang.Seqable bindings)]
+                       ;; Type hint to Indexed vs MapEntry just because MapEntry seems to be a
+                       ;; less stable impl detail to rely on.
+                       (loop [^clojure.lang.Indexed entry (.first entries)
+                              entries (.next entries)]
+                         (let [k (some-> entry (.nth 0))
+                               v (some-> entry (.nth 1))]
+                           (if (and k
+                                    ;; other-bindings will always be persistent map so invoke
+                                    ;; it directly.  It is faster than `get`.
+                                    (compare-fn v (other-bindings k)))
+                             (recur (some-> entries .first)
+                                    (some-> entries .next))
+                             ;; If there is no k left, then every entry matched.  If there is a k,
+                             ;; that means the comparison failed, so the maps aren't equal.
+                             (not k))))
+                       ;; Empty bindings on both sides.
+                       true))) 
+
+              ;; Check the :matches on each token.  :matches need to be in the same order on both
+              ;; tokens to be considered the same.
+              (let [^clojure.lang.Indexed matches (:matches token)
+                    ^clojure.lang.Indexed other-matches (:matches other-token)
+                    count-matches (.size ^java.util.List matches)]
+                (and (= count-matches
+                        (.size ^java.util.List other-matches))
+                     (loop [i 0]
+                       (cond
+                         (= i count-matches)
+                         true
+
+                         ;; Compare node-id's first.  Fallback to comparing facts.  This will
+                         ;; most very likely be the most expensive part to execute.
+                         (let [^clojure.lang.Indexed m (.nth matches i)
+                               ^clojure.lang.Indexed om (.nth other-matches i)]
+                           ;; A token :matches tuple is of the form [fact node-id].
+                           (and (= (.nth m 1) (.nth om 1))
+                                (compare-fn (.nth m 0) (.nth om 0))))
+                         (recur (inc i))
+
+                         :else
+                         false))))))))
+
+#?(:clj
+   (defprotocol IdentityComparable
+     (using-token-identity [this bool])))
+
+#?(:clj
+   (deftype RuleOrderedActivation [node-id
+                                   token
+                                   activation
+                                   rule-load-order
+                                   ^:unsynchronized-mutable use-token-identity?]
```

Comment:
So I *think* this is safe, since the mutation is only set to false and then under specific circumstances. Sessions can be shared between threads, of course, and mutations to this won't be visible to other threads since it's not at least volatile. But that should also be safe, since the calling thread would just set this to false again.

Might be worth adding a comment here as to these constraints.